File: 40. Combination Sum II/40. Combination Sum II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
  candidates.sort((a, b) => a - b);
  const res = [];
  dfs(candidates, target, 0, [], res, new Array(candidates.length));
  return res;
};
var dfs = function(nums, remain, index, cur, res, used) {
  if (remain < 0) return;
  if (remain == 0) res.push(cur.slice());

  for (let i = index; i < nums.length; i++) {
    if (i > 0 && !used[i - 1] && nums[i - 1] == nums[i]) continue;
    used[i] = true;
    cur.push(nums[i]);
    dfs(nums, remain - nums[i], i + 1, cur, res, used);
    used[i] = false;
    cur.pop(nums[i]);
  }
};

File: 40. Combination Sum II/40. Combination Sum II.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list =  new ArrayList<>();
        getComb2(candidates,target, res, list, 0);
        return res;       
    }
    
    private void getComb2(int[]nums, int target, List<List<Integer>> res, List<Integer> current, int pos){
        if(target < 0) return;
        else if(target == 0) res.add(new ArrayList<>(current));
        else{
            for(int i = pos; i < nums.length; i++){      
                current.add(nums[i]);
                getComb2(nums,target - nums[i], res, current, i + 1);
                current.remove(current.size() - 1);
                while(i < nums.length - 1 && nums[i] == nums[i + 1]){
                        i++;
                }
            }
        }        
    }
}